// @0xgasless/smart-account/dist/_cjs/modules/session-storage/SessionFileStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionFileStorage = void 0;
const viem_1 = require("viem");
const accounts_1 = require("viem/accounts");
const __1 = require("../../index.js");
const account_1 = require("../../account/index.js");
class SessionFileStorage {
    constructor(smartAccountAddress) {
        Object.defineProperty(this, "smartAccountAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.smartAccountAddress = smartAccountAddress.toLowerCase();
    }
    async readDataFromFile(type) {
        return new Promise((resolve) => {
            fs.readFile(this.getStorageFilePath(type), "utf8", (err, data) => {
                if (err) {
                    resolve(undefined);
                }
                else {
                    if (!data) {
                        resolve(null);
                    }
                    else {
                        resolve(JSON.parse(data));
                    }
                }
            });
        });
    }
    getStorageFilePath(type) {
        return `${__dirname}/sessionStorageData/${this.smartAccountAddress}_${type}.json`;
    }
    async writeDataToFile(data, type) {
        console.log("");
        return new Promise((resolve, reject) => {
            const filePath = this.getStorageFilePath(type);
            fs.writeFile(filePath, JSON.stringify(data), "utf8", (err) => {
                if (err) {
                    console.log({ err }, JSON.stringify(data));
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    validateSearchParam(param) {
        if (param.sessionID) {
            return;
        }
        if (!param.sessionID &&
            param.sessionPublicKey &&
            param.sessionValidationModule) {
            return;
        }
        throw new Error("Either pass sessionId or a combination of sessionPublicKey and sessionValidationModule address.");
    }
    async getSessionStore() {
        try {
            const data = await this.readDataFromFile("sessions");
            return data || { merkleRoot: "", leafNodes: [] };
        }
        catch (error) {
            console.log({ error });
        }
    }
    async getSignerStore() {
        try {
            const data = await this.readDataFromFile("signers");
            return data || {};
        }
        catch (error) {
            console.log({ error });
        }
    }
    toLowercaseAddress(address) {
        return address.toLowerCase();
    }
    async getSessionData(param) {
        const sessions = (await this.getSessionStore()).leafNodes;
        const session = sessions.find((s) => {
            if (param.sessionID) {
                return (s.sessionID === param.sessionID &&
                    (!param.status || s.status === param.status));
            }
            if (param.sessionPublicKey && param.sessionValidationModule) {
                return (s.sessionPublicKey ===
                    this.toLowercaseAddress(param.sessionPublicKey) &&
                    s.sessionValidationModule ===
                        this.toLowercaseAddress(param.sessionValidationModule) &&
                    (!param.status || s.status === param.status));
            }
            return undefined;
        });
        if (!session) {
            throw new Error("Session not found.");
        }
        return session;
    }
    async addSessionData(leaf) {
        account_1.Logger.log("Add session Data");
        const data = await this.getSessionStore();
        leaf.sessionValidationModule = this.toLowercaseAddress(leaf.sessionValidationModule);
        leaf.sessionPublicKey = this.toLowercaseAddress(leaf.sessionPublicKey);
        data.leafNodes.push(leaf);
        await this.writeDataToFile(data, "sessions");
    }
    async updateSessionStatus(param, status) {
        this.validateSearchParam(param);
        const data = await this.getSessionStore();
        const session = data.leafNodes.find((s) => {
            if (param.sessionID) {
                return s.sessionID === param.sessionID;
            }
            if (param.sessionPublicKey && param.sessionValidationModule) {
                return (s.sessionPublicKey ===
                    this.toLowercaseAddress(param.sessionPublicKey) &&
                    s.sessionValidationModule ===
                        this.toLowercaseAddress(param.sessionValidationModule));
            }
            return undefined;
        });
        if (!session) {
            throw new Error("Session not found.");
        }
        session.status = status;
        await this.writeDataToFile(data, "sessions");
    }
    async clearPendingSessions() {
        const data = await this.getSessionStore();
        data.leafNodes = data.leafNodes.filter((s) => s.status !== "PENDING");
        await this.writeDataToFile(data, "sessions");
    }
    async addSigner(signerData, chain) {
        const signers = await this.getSignerStore();
        const signer = signerData ?? (0, __1.getRandomSigner)();
        const accountSigner = (0, accounts_1.privateKeyToAccount)(signer.pvKey);
        const client = (0, viem_1.createWalletClient)({
            account: accountSigner,
            chain,
            transport: (0, viem_1.http)()
        });
        const walletClientSigner = new account_1.WalletClientSigner(client, "json-rpc");
        signers[this.toLowercaseAddress(accountSigner.address)] = signer;
        await this.writeDataToFile(signers, "signers");
        return walletClientSigner;
    }
    async getSignerByKey(sessionPublicKey, chain) {
        const signers = await this.getSignerStore();
        account_1.Logger.log("Got signers", signers);
        const signerData = signers[this.toLowercaseAddress(sessionPublicKey)];
        if (!signerData) {
            throw new Error("Signer not found.");
        }
        account_1.Logger.log(signerData.pvKey, "PVKEY");
        const signer = (0, accounts_1.privateKeyToAccount)(signerData.pvKey);
        const walletClient = (0, viem_1.createWalletClient)({
            account: signer,
            chain,
            transport: (0, viem_1.http)()
        });
        return new account_1.WalletClientSigner(walletClient, "json-rpc");
    }
    async getSignerBySession(param, chain) {
        const session = await this.getSessionData(param);
        account_1.Logger.log("got session");
        const signer = await this.getSignerByKey(session.sessionPublicKey, chain);
        return signer;
    }
    async getAllSessionData(param) {
        const sessions = (await this.getSessionStore()).leafNodes;
        if (!param || !param.status) {
            return sessions;
        }
        return sessions.filter((s) => s.status === param.status);
    }
    async getMerkleRoot() {
        return (await this.getSessionStore()).merkleRoot;
    }
    async setMerkleRoot(merkleRoot) {
        const data = await this.getSessionStore();
        data.merkleRoot = merkleRoot;
        await this.writeDataToFile(data, "sessions");
    }
}
exports.SessionFileStorage = SessionFileStorage;
