// @0xgasless/agentkit/dist/agentkit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Agentkit = void 0;
const viem_1 = require("viem");
const accounts_1 = require("viem/accounts");
const smart_account_1 = require("@0xgasless/smart-account");
const constants_1 = require("./constants");
class Agentkit {
    constructor(config) {
        if (!constants_1.supportedChains[config.chainID]) {
            throw new Error(`Chain ID ${config.chainID} is not supported`);
        }
    }
    static async configureWithWallet(config) {
        if (!config.apiKey || config.apiKey === "") {
            throw new Error("API_KEY is required for smart agent configuration");
        }
        const agentkit = new Agentkit(config);
        try {
            let account;
            if (config.privateKey) {
                account = (0, accounts_1.privateKeyToAccount)(config.privateKey);
            }
            else if (config.mnemonicPhrase) {
                account = (0, accounts_1.mnemonicToAccount)(config.mnemonicPhrase, {
                    accountIndex: config.accountPath || 0,
                });
            }
            else {
                throw new Error("Either privateKey or mnemonicPhrase must be provided");
            }
            // Create wallet client
            const wallet = (0, viem_1.createWalletClient)({
                account,
                chain: constants_1.supportedChains[config.chainID],
                transport: config.rpcUrl ? (0, viem_1.http)(config.rpcUrl) : (0, viem_1.http)(),
            });
            // Configure smart account
            const bundlerUrl = `https://bundler.0xgasless.com/${config.chainID}`;
            const paymasterUrl = `https://paymaster.0xgasless.com/v1/${config.chainID}/rpc/${config.apiKey}`;
            agentkit.smartAccount = await (0, smart_account_1.createSmartAccountClient)({
                bundlerUrl,
                paymasterUrl,
                chainId: config.chainID,
                signer: wallet,
            });
        }
        catch (error) {
            throw new Error(`Failed to initialize smart account: ${error}`);
        }
        return agentkit;
    }
    async run(action, args) {
        if (!this.smartAccount) {
            return `Unable to run Action: ${action.name}. A Smart Account is required. Please configure Agentkit with a Wallet to run this action.`;
        }
        return await action.func(this.smartAccount, args);
    }
    async getAddress() {
        if (!this.smartAccount) {
            throw new Error("Smart account not configured");
        }
        return await this.smartAccount.getAddress();
    }
    async getChainId() {
        if (!this.smartAccount) {
            throw new Error("Smart account not configured");
        }
        return this.smartAccount.SmartAccountConfig.chainId;
    }
}
exports.Agentkit = Agentkit;
