// @0xgasless/smart-account/dist/_esm/modules/session-storage/SessionFileStorage.js
import { http, createWalletClient } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { getRandomSigner } from "../../index.js";
import { Logger, WalletClientSigner } from "../../account/index.js";
export class SessionFileStorage {
    constructor(smartAccountAddress) {
        Object.defineProperty(this, "smartAccountAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.smartAccountAddress = smartAccountAddress.toLowerCase();
    }
    // This method reads data from the file and returns it in the JSON format
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    async readDataFromFile(type) {
        return new Promise((resolve) => {
            // @ts-ignore
            fs.readFile(this.getStorageFilePath(type), "utf8", (err, data) => {
                if (err) {
                    // Handle errors appropriately
                    resolve(undefined);
                }
                else {
                    if (!data) {
                        resolve(null);
                    }
                    else {
                        resolve(JSON.parse(data));
                    }
                    //   resolve(JSON.parse(data));
                }
            });
        });
    }
    getStorageFilePath(type) {
        return `${__dirname}/sessionStorageData/${this.smartAccountAddress}_${type}.json`;
    }
    async writeDataToFile(
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    data, type) {
        console.log("");
        return new Promise((resolve, reject) => {
            const filePath = this.getStorageFilePath(type);
            // @ts-ignore
            fs.writeFile(filePath, JSON.stringify(data), "utf8", (err) => {
                if (err) {
                    // Handle errors appropriately
                    console.log({ err }, JSON.stringify(data));
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    validateSearchParam(param) {
        if (param.sessionID) {
            return;
        }
        if (!param.sessionID &&
            param.sessionPublicKey &&
            param.sessionValidationModule) {
            return;
        }
        throw new Error("Either pass sessionId or a combination of sessionPublicKey and sessionValidationModule address.");
    }
    // Session store is in the form of mekrleRoot and leafnodes, each object will have a root and an array of leafNodes.
    async getSessionStore() {
        try {
            const data = await this.readDataFromFile("sessions");
            return data || { merkleRoot: "", leafNodes: [] };
        }
        catch (error) {
            // Handle errors appropriately
            console.log({ error });
        }
    }
    async getSignerStore() {
        try {
            const data = await this.readDataFromFile("signers");
            return data || {};
        }
        catch (error) {
            console.log({ error });
            // Handle errors appropriately
        }
    }
    // private getStorageKey(type: "sessions" | "signers"): string {
    //   return `${this.smartAccountAddress}_${type}`
    // }
    toLowercaseAddress(address) {
        return address.toLowerCase();
    }
    async getSessionData(param) {
        const sessions = (await this.getSessionStore()).leafNodes;
        const session = sessions.find((s) => {
            if (param.sessionID) {
                return (s.sessionID === param.sessionID &&
                    (!param.status || s.status === param.status));
            }
            if (param.sessionPublicKey && param.sessionValidationModule) {
                return (s.sessionPublicKey ===
                    this.toLowercaseAddress(param.sessionPublicKey) &&
                    s.sessionValidationModule ===
                        this.toLowercaseAddress(param.sessionValidationModule) &&
                    (!param.status || s.status === param.status));
            }
            return undefined;
        });
        if (!session) {
            throw new Error("Session not found.");
        }
        return session;
    }
    async addSessionData(leaf) {
        Logger.log("Add session Data");
        const data = await this.getSessionStore();
        leaf.sessionValidationModule = this.toLowercaseAddress(leaf.sessionValidationModule);
        leaf.sessionPublicKey = this.toLowercaseAddress(leaf.sessionPublicKey);
        data.leafNodes.push(leaf);
        await this.writeDataToFile(data, "sessions"); // Use 'sessions' as the type
    }
    async updateSessionStatus(param, status) {
        this.validateSearchParam(param);
        const data = await this.getSessionStore();
        const session = data.leafNodes.find((s) => {
            if (param.sessionID) {
                return s.sessionID === param.sessionID;
            }
            if (param.sessionPublicKey && param.sessionValidationModule) {
                return (s.sessionPublicKey ===
                    this.toLowercaseAddress(param.sessionPublicKey) &&
                    s.sessionValidationModule ===
                        this.toLowercaseAddress(param.sessionValidationModule));
            }
            return undefined;
        });
        if (!session) {
            throw new Error("Session not found.");
        }
        session.status = status;
        await this.writeDataToFile(data, "sessions"); // Use 'sessions' as the type
    }
    async clearPendingSessions() {
        const data = await this.getSessionStore();
        data.leafNodes = data.leafNodes.filter((s) => s.status !== "PENDING");
        await this.writeDataToFile(data, "sessions"); // Use 'sessions' as the type
    }
    async addSigner(signerData, chain) {
        const signers = await this.getSignerStore();
        const signer = signerData ?? getRandomSigner();
        const accountSigner = privateKeyToAccount(signer.pvKey);
        const client = createWalletClient({
            account: accountSigner,
            chain,
            transport: http()
        });
        const walletClientSigner = new WalletClientSigner(client, "json-rpc" // signerType
        );
        signers[this.toLowercaseAddress(accountSigner.address)] = signer;
        await this.writeDataToFile(signers, "signers"); // Use 'signers' as the type
        return walletClientSigner;
    }
    async getSignerByKey(sessionPublicKey, chain) {
        const signers = await this.getSignerStore();
        Logger.log("Got signers", signers);
        const signerData = signers[this.toLowercaseAddress(sessionPublicKey)];
        if (!signerData) {
            throw new Error("Signer not found.");
        }
        Logger.log(signerData.pvKey, "PVKEY");
        const signer = privateKeyToAccount(signerData.pvKey);
        const walletClient = createWalletClient({
            account: signer,
            chain,
            transport: http()
        });
        return new WalletClientSigner(walletClient, "json-rpc");
    }
    async getSignerBySession(param, chain) {
        const session = await this.getSessionData(param);
        Logger.log("got session");
        const signer = await this.getSignerByKey(session.sessionPublicKey, chain);
        return signer;
    }
    async getAllSessionData(param) {
        const sessions = (await this.getSessionStore()).leafNodes;
        if (!param || !param.status) {
            return sessions;
        }
        return sessions.filter((s) => s.status === param.status);
    }
    async getMerkleRoot() {
        return (await this.getSessionStore()).merkleRoot;
    }
    async setMerkleRoot(merkleRoot) {
        const data = await this.getSessionStore();
        data.merkleRoot = merkleRoot;
        await this.writeDataToFile(data, "sessions"); // Use 'sessions' as the type
    }
}
