// @0xgasless/smart-account/dist/_esm/modules/utils/Constants.js
export const DEFAULT_MODULE_VERSION = "V1_0_0";
export const DEFAULT_ECDSA_OWNERSHIP_MODULE = "0x024F01e9De3D0258DF7930f146F13D4eF5fe0750";
export const ECDSA_OWNERSHIP_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x024F01e9De3D0258DF7930f146F13D4eF5fe0750"
};
export const DEFAULT_SESSION_KEY_MANAGER_MODULE = "0xB8E4C94C5cb050EFcEB13b8e1Af7A12Ed1B14AE8";
export const SESSION_MANAGER_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x000000456b395c4e107e0302553B90D1eF4a32e9",
    V1_0_1: "0xB8E4C94C5cb050EFcEB13b8e1Af7A12Ed1B14AE8"
};
export const DEFAULT_BATCHED_SESSION_ROUTER_MODULE = "0x00000D09967410f8C76752A104c9848b57ebba55";
export const BATCHED_SESSION_ROUTER_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x00000D09967410f8C76752A104c9848b57ebba55"
};
export const DEFAULT_ERC20_MODULE = "0xD282F13f9E67DF27CE856771C84F2E15971a8bD8";
export const DEFAULT_MULTICHAIN_MODULE = "0x5E923255DAc3d1240513Ecd67Dd6F1E44c6F6dFe";
export const MULTICHAIN_VALIDATION_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x5E923255DAc3d1240513Ecd67Dd6F1E44c6F6dFe"
};
export const DEFAULT_ABI_SVM_MODULE = "0x000006bC2eCdAe38113929293d241Cf252D91861";
export const MODULE_ADDRESSES = {
    ERC20: DEFAULT_ERC20_MODULE,
    ABI: DEFAULT_ABI_SVM_MODULE
};
