// @0xgasless/smart-account/dist/_esm/modules/utils/Helper.js
import { encodeAbiParameters, keccak256, parseAbiParameters } from "viem";
import { generatePrivateKey, privateKeyToAccount } from "viem/accounts";
import { getChain } from "../../account/index.js";
function packUserOp(op, forSignature = true) {
    if (!op.initCode || !op.callData || !op.paymasterAndData)
        throw new Error("Missing userOp properties");
    if (forSignature) {
        return encodeAbiParameters(parseAbiParameters("address, uint256, bytes32, bytes32, uint256, uint256, uint256, uint256, uint256, bytes32"), [
            op.sender,
            BigInt(op.nonce),
            keccak256(op.initCode),
            keccak256(op.callData),
            BigInt(op.callGasLimit),
            BigInt(op.verificationGasLimit),
            BigInt(op.preVerificationGas),
            BigInt(op.maxFeePerGas),
            BigInt(op.maxPriorityFeePerGas),
            keccak256(op.paymasterAndData)
        ]);
    }
    // for the purpose of calculating gas cost encode also signature (and no keccak of bytes)
    return encodeAbiParameters(parseAbiParameters("address, uint256, bytes, bytes, uint256, uint256, uint256, uint256, uint256, bytes, bytes"), [
        op.sender,
        BigInt(op.nonce),
        op.initCode,
        op.callData,
        BigInt(op.callGasLimit),
        BigInt(op.verificationGasLimit),
        BigInt(op.preVerificationGas),
        BigInt(op.maxFeePerGas),
        BigInt(op.maxPriorityFeePerGas),
        op.paymasterAndData,
        op.signature
    ]);
}
export const getUserOpHash = (userOp, entryPointAddress, chainId) => {
    const userOpHash = keccak256(packUserOp(userOp, true));
    const enc = encodeAbiParameters(parseAbiParameters("bytes32, address, uint256"), [userOpHash, entryPointAddress, BigInt(chainId)]);
    return keccak256(enc);
};
export const getRandomSigner = () => {
    const pkey = generatePrivateKey();
    const account = privateKeyToAccount(pkey);
    return {
        pvKey: pkey,
        pbKey: account.address
    };
};
export const parseChain = (chainInfo) => {
    if (typeof chainInfo === "number")
        return getChain(chainInfo);
    return chainInfo;
};
