// @0xgasless/smart-account/dist/_cjs/modules/utils/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODULE_ADDRESSES = exports.DEFAULT_ABI_SVM_MODULE = exports.MULTICHAIN_VALIDATION_MODULE_ADDRESSES_BY_VERSION = exports.DEFAULT_MULTICHAIN_MODULE = exports.DEFAULT_ERC20_MODULE = exports.BATCHED_SESSION_ROUTER_MODULE_ADDRESSES_BY_VERSION = exports.DEFAULT_BATCHED_SESSION_ROUTER_MODULE = exports.SESSION_MANAGER_MODULE_ADDRESSES_BY_VERSION = exports.DEFAULT_SESSION_KEY_MANAGER_MODULE = exports.ECDSA_OWNERSHIP_MODULE_ADDRESSES_BY_VERSION = exports.DEFAULT_ECDSA_OWNERSHIP_MODULE = exports.DEFAULT_MODULE_VERSION = void 0;
exports.DEFAULT_MODULE_VERSION = "V1_0_0";
exports.DEFAULT_ECDSA_OWNERSHIP_MODULE = "0x024F01e9De3D0258DF7930f146F13D4eF5fe0750";
exports.ECDSA_OWNERSHIP_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x024F01e9De3D0258DF7930f146F13D4eF5fe0750"
};
exports.DEFAULT_SESSION_KEY_MANAGER_MODULE = "0xB8E4C94C5cb050EFcEB13b8e1Af7A12Ed1B14AE8";
exports.SESSION_MANAGER_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x000000456b395c4e107e0302553B90D1eF4a32e9",
    V1_0_1: "0xB8E4C94C5cb050EFcEB13b8e1Af7A12Ed1B14AE8"
};
exports.DEFAULT_BATCHED_SESSION_ROUTER_MODULE = "0x00000D09967410f8C76752A104c9848b57ebba55";
exports.BATCHED_SESSION_ROUTER_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x00000D09967410f8C76752A104c9848b57ebba55"
};
exports.DEFAULT_ERC20_MODULE = "0xD282F13f9E67DF27CE856771C84F2E15971a8bD8";
exports.DEFAULT_MULTICHAIN_MODULE = "0x5E923255DAc3d1240513Ecd67Dd6F1E44c6F6dFe";
exports.MULTICHAIN_VALIDATION_MODULE_ADDRESSES_BY_VERSION = {
    V1_0_0: "0x5E923255DAc3d1240513Ecd67Dd6F1E44c6F6dFe"
};
exports.DEFAULT_ABI_SVM_MODULE = "0x000006bC2eCdAe38113929293d241Cf252D91861";
exports.MODULE_ADDRESSES = {
    ERC20: exports.DEFAULT_ERC20_MODULE,
    ABI: exports.DEFAULT_ABI_SVM_MODULE
};
