// @0xgasless/agentkit/dist/services/keyManagementService.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.storeKey = storeKey;
exports.getStoredPrivateKeyByAddress = getStoredPrivateKeyByAddress;
exports.generateNewViemPrivateKey = generateNewViemPrivateKey;
exports.getAddressFromPrivateKey = getAddressFromPrivateKey;
const accounts_1 = require("viem/accounts");
const path_1 = __importDefault(require("path"));
// import { EOL } from 'os'; // EOL not strictly needed for console.log, removing for now
// Determine if running in a Node.js-like environment
const isNodeLikeEnvironment = typeof process !== "undefined" && process.versions != null && process.versions.node != null;
// Dynamically imported sqlite3 types and instance
let sqlite3Module = null; // To hold the imported sqlite3 module itself
let SqliteDatabaseConstructor = null; // To hold the Database constructor from the module
// Database path: agentkit/keys.db (workspace root)
const DB_FILE_PATH = path_1.default.resolve(process.cwd(), "keys.db");
// biome-ignore lint/suspicious/noExplicitAny: <explanation>
let dbInstanceInternal = null; // Will be an instance of SqliteDatabaseConstructor in Node
// biome-ignore lint/suspicious/noExplicitAny: <explanation>
let dbInitializationPromiseInternal = null;
async function loadSqlite3IfNotLoaded() {
    if (!isNodeLikeEnvironment || sqlite3Module)
        return;
    try {
        const importedModule = await Promise.resolve().then(() => __importStar(require("sqlite3")));
        sqlite3Module = importedModule.default || importedModule;
        if (typeof sqlite3Module === "object" &&
            sqlite3Module !== null &&
            "Database" in sqlite3Module) {
            SqliteDatabaseConstructor = sqlite3Module.Database;
        }
        else {
            console.warn("SQLite3 Database constructor not found on the imported module.");
            SqliteDatabaseConstructor = null;
            sqlite3Module = null;
        }
        if (SqliteDatabaseConstructor) {
            console.log("SQLite3 module and Database constructor loaded successfully.");
        }
        else if (sqlite3Module) {
            console.warn("SQLite3 module loaded, but Database constructor could not be identified.");
        }
    }
    catch (e) {
        console.warn("Failed to load sqlite3 module. Key storage will be disabled.", e instanceof Error ? e.message : String(e));
        sqlite3Module = null;
        SqliteDatabaseConstructor = null;
    }
}
// Initialize SQLite and ensure table exists (Node.js only)
// biome-ignore lint/suspicious/noExplicitAny: <explanation>
async function getKeystoreDB() {
    // Returns Promise<sqlite3.Database | null>
    if (!isNodeLikeEnvironment) {
        return null;
    }
    if (!sqlite3Module || !SqliteDatabaseConstructor) {
        await loadSqlite3IfNotLoaded();
        if (!sqlite3Module || !SqliteDatabaseConstructor) {
            console.warn("SQLite3 not available after load attempt. Cannot get DB instance.");
            return null;
        }
    }
    if (dbInstanceInternal) {
        return Promise.resolve(dbInstanceInternal);
    }
    if (dbInitializationPromiseInternal) {
        return dbInitializationPromiseInternal;
    }
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    dbInitializationPromiseInternal = new Promise((resolve, reject) => {
        // Type assertion for the constructor
        const DbConstructor = SqliteDatabaseConstructor;
        if (!DbConstructor) {
            const err = new Error("SQLite Database constructor is not available.");
            console.error(err.message);
            dbInitializationPromiseInternal = null;
            return reject(err);
        }
        try {
            const newDb = new DbConstructor(DB_FILE_PATH, (err) => {
                if (err) {
                    console.error(`Error opening database at ${DB_FILE_PATH}:`, err.message);
                    dbInitializationPromiseInternal = null;
                    return reject(err);
                }
                console.log(`Connected to the SQLite database: ${DB_FILE_PATH}`);
                newDb.run(`CREATE TABLE IF NOT EXISTS keystore (
          address TEXT PRIMARY KEY,
          private_key TEXT NOT NULL
        )`, (createTableErr) => {
                    if (createTableErr) {
                        console.error("Error creating keystore table:", createTableErr.message);
                        newDb.close((closeErr) => {
                            if (closeErr)
                                console.error("Error closing DB after table creation failure:", closeErr.message);
                        });
                        dbInitializationPromiseInternal = null;
                        return reject(createTableErr);
                    }
                    console.log("Keystore table created or already exists.");
                    dbInstanceInternal = newDb;
                    resolve(dbInstanceInternal);
                });
            });
        }
        catch (constructorError) {
            // Catch errors from `new DbConstructor` itself if it's not a valid constructor
            console.error("Error instantiating SQLite Database:", constructorError);
            dbInitializationPromiseInternal = null;
            reject(constructorError);
        }
    }).catch(err => {
        dbInitializationPromiseInternal = null;
        throw err;
    });
    return dbInitializationPromiseInternal;
}
// Store private key (Node.js only)
async function storeKey(address, privateKey) {
    if (!isNodeLikeEnvironment)
        return false;
    const db = await getKeystoreDB();
    if (!db) {
        console.warn("DB not available, cannot store key.");
        return false;
    }
    return new Promise((resolve, reject) => {
        const sql = `INSERT OR REPLACE INTO keystore (address, private_key) VALUES (?, ?)`;
        db.run(sql, [address, privateKey], function (err) {
            if (err) {
                console.error("Error storing key:", err.message);
                return reject(err);
            }
            console.log(`Key for address ${address} stored/updated. Rows: ${this.changes}`);
            resolve(true);
        });
    });
}
// Retrieve stored private key by address (Node.js only)
async function getStoredPrivateKeyByAddress(address) {
    if (!isNodeLikeEnvironment)
        return null;
    const db = await getKeystoreDB();
    if (!db)
        return null;
    return new Promise((resolve, reject) => {
        const sql = `SELECT private_key FROM keystore WHERE address = ?`;
        db.get(sql, [address], (err, row) => {
            if (err) {
                console.error("Error retrieving key:", err.message);
                return reject(err);
            }
            resolve(row?.private_key || null);
        });
    });
}
// Generate a new private key using viem
function generateNewViemPrivateKey() {
    return (0, accounts_1.generatePrivateKey)();
}
// Get address from a private key using viem
function getAddressFromPrivateKey(privateKey) {
    const account = (0, accounts_1.privateKeyToAccount)(privateKey);
    return account.address;
}
// Attempt to load sqlite3 when service is initialized in Node.js environment
if (isNodeLikeEnvironment) {
    // Fire and forget, errors handled internally
    loadSqlite3IfNotLoaded().catch(err => {
        console.warn("Initial attempt to load sqlite3 failed on module load:", err instanceof Error ? err.message : String(err));
    });
}
