// @0xgasless/agentkit/dist/BaseActions/FomatHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatHelperActions = exports.FromBytesAction = exports.FromBytesPrompt = exports.FromBytesSchema = exports.ToBytesAction = exports.ToBytesPrompt = exports.ToBytesSchema = exports.FromHexAction = exports.FromHexPrompt = exports.FromHexSchema = exports.ToHexAction = exports.ToHexPrompt = exports.ToHexSchema = exports.ParseEtherAction = exports.ParseEtherPrompt = exports.ParseEtherSchema = exports.FormatEtherAction = exports.FormatEtherPrompt = exports.FormatEtherSchema = exports.ParseUnitsAction = exports.ParseUnitsPrompt = exports.ParseUnitsSchema = exports.FormatUnitsAction = exports.FormatUnitsPrompt = exports.FormatUnitsSchema = void 0;
exports.formatUnitsFunc = formatUnitsFunc;
exports.parseUnitsFunc = parseUnitsFunc;
exports.formatEtherFunc = formatEtherFunc;
exports.parseEtherFunc = parseEtherFunc;
exports.toHexFunc = toHexFunc;
exports.fromHexFunc = fromHexFunc;
exports.toBytesFunc = toBytesFunc;
exports.fromBytesFunc = fromBytesFunc;
const zod_1 = require("zod");
const viem_1 = require("viem");
// Common error handling function
function formatError(toolName, error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`Error in ${toolName}:`, error);
    return `Error in ${toolName}: ${message}`;
}
// --- 1. Format Units ---
exports.FormatUnitsSchema = zod_1.z.object({
    value: zod_1.z.string().regex(/^-?\d+$/, "Value must be a string representing an integer (wei)."),
    decimals: zod_1.z.number().int().positive("Decimals must be a positive integer."),
});
exports.FormatUnitsPrompt = `
Name: format_units
Description: Converts a numerical value (as a string in wei or smallest unit) into a human-readable string representation based on the specified number of decimals.
Usage: Useful for displaying token balances or amounts stored as large integers in a user-friendly format (e.g., converting wei to Ether, or smallest token unit to standard unit).
Input Parameters:
  - value (string, required): The numerical value in its smallest unit (e.g., wei) as a string.
  - decimals (number, required): The number of decimal places the unit uses (e.g., 18 for Ether, 6 for USDC).
Output:
  - On success: Returns the formatted string. Example: "1.2345"
  - On failure: Returns an error message.
`;
async function formatUnitsFunc(_wallet, { value, decimals }) {
    try {
        const formatted = (0, viem_1.formatUnits)(BigInt(value), decimals);
        return `Formatted Value: ${formatted}`;
    }
    catch (error) {
        return formatError("format_units", error);
    }
}
class FormatUnitsAction {
    constructor() {
        this.name = "format_units";
        this.description = exports.FormatUnitsPrompt;
        this.argsSchema = exports.FormatUnitsSchema;
        this.func = formatUnitsFunc;
        this.smartAccountRequired = false;
    }
}
exports.FormatUnitsAction = FormatUnitsAction;
// --- 2. Parse Units ---
exports.ParseUnitsSchema = zod_1.z.object({
    value: zod_1.z
        .string()
        .regex(/^-?(\d+(\.\d+)?|\.\d+)$/, "Value must be a string representing a decimal number."),
    decimals: zod_1.z.number().int().positive("Decimals must be a positive integer."),
});
exports.ParseUnitsPrompt = `
Name: parse_units
Description: Converts a human-readable string representation of a value (e.g., "1.23") into its integer representation (e.g., wei) based on the specified number of decimals.
Usage: Useful for converting user input (like "0.5" ETH) into the format required for blockchain transactions (wei).
Input Parameters:
  - value (string, required): The human-readable decimal value as a string.
  - decimals (number, required): The number of decimal places the unit uses (e.g., 18 for Ether, 6 for USDC).
Output:
  - On success: Returns the parsed value as a string integer. Example: "1230000000000000000"
  - On failure: Returns an error message.
`;
async function parseUnitsFunc(_wallet, { value, decimals }) {
    try {
        const parsed = (0, viem_1.parseUnits)(value, decimals);
        return `Parsed Value (Wei): ${parsed.toString()}`;
    }
    catch (error) {
        return formatError("parse_units", error);
    }
}
class ParseUnitsAction {
    constructor() {
        this.name = "parse_units";
        this.description = exports.ParseUnitsPrompt;
        this.argsSchema = exports.ParseUnitsSchema;
        this.func = parseUnitsFunc;
        this.smartAccountRequired = false;
    }
}
exports.ParseUnitsAction = ParseUnitsAction;
// --- 3. Format Ether ---
exports.FormatEtherSchema = zod_1.z.object({
    value: zod_1.z.string().regex(/^-?\d+$/, "Value must be a string representing an integer (wei)."),
});
exports.FormatEtherPrompt = `
Name: format_ether
Description: Converts a value in wei (as a string) into its Ether representation (string). Shortcut for format_units with decimals=18.
Usage: Specifically for converting native currency wei values to Ether.
Input Parameters:
  - value (string, required): The value in wei as a string.
Output:
  - On success: Returns the formatted Ether string. Example: "1.2345"
  - On failure: Returns an error message.
`;
async function formatEtherFunc(_wallet, { value }) {
    try {
        const formatted = (0, viem_1.formatEther)(BigInt(value));
        return `Formatted Ether: ${formatted}`;
    }
    catch (error) {
        return formatError("format_ether", error);
    }
}
class FormatEtherAction {
    constructor() {
        this.name = "format_ether";
        this.description = exports.FormatEtherPrompt;
        this.argsSchema = exports.FormatEtherSchema;
        this.func = formatEtherFunc;
        this.smartAccountRequired = false;
    }
}
exports.FormatEtherAction = FormatEtherAction;
// --- 4. Parse Ether ---
exports.ParseEtherSchema = zod_1.z.object({
    value: zod_1.z
        .string()
        .regex(/^-?(\d+(\.\d+)?|\.\d+)$/, "Value must be a string representing a decimal number."),
});
exports.ParseEtherPrompt = `
Name: parse_ether
Description: Converts a string representation of Ether into its wei representation (as a string). Shortcut for parse_units with decimals=18.
Usage: Specifically for converting Ether values (like user input) to wei for transactions.
Input Parameters:
  - value (string, required): The Ether value as a string.
Output:
  - On success: Returns the parsed value in wei as a string. Example: "1234500000000000000"
  - On failure: Returns an error message.
`;
async function parseEtherFunc(_wallet, { value }) {
    try {
        const parsed = (0, viem_1.parseEther)(value);
        return `Parsed Wei: ${parsed.toString()}`;
    }
    catch (error) {
        return formatError("parse_ether", error);
    }
}
class ParseEtherAction {
    constructor() {
        this.name = "parse_ether";
        this.description = exports.ParseEtherPrompt;
        this.argsSchema = exports.ParseEtherSchema;
        this.func = parseEtherFunc;
        this.smartAccountRequired = false;
    }
}
exports.ParseEtherAction = ParseEtherAction;
// --- 5. To Hex ---
exports.ToHexSchema = zod_1.z.object({
    value: zod_1.z
        .union([
        zod_1.z.string(),
        zod_1.z.number(),
        zod_1.z.bigint(),
        zod_1.z.boolean(),
        // z.instanceof(Uint8Array) // Difficult to represent Uint8Array as string input reliably
    ])
        .describe("The value to convert to Hex. Can be string, number, bigint, or boolean."),
    // size: z.number().int().positive().optional().describe("The desired byte size of the hex value.")
});
exports.ToHexPrompt = `
Name: to_hex
Description: Converts a string, number, bigint, or boolean value into a hexadecimal string (0x...).
Usage: Useful for converting various data types into the hex format commonly used in Ethereum. Note: Cannot directly convert byte arrays via agent input, use other tools first if needed.
Input Parameters:
  - value (string | number | bigint | boolean, required): The value to encode.
Output:
  - On success: Returns the hex string. Example: "Hex: 0x60" (for number 96)
  - On failure: Returns an error message.
`;
async function toHexFunc(_wallet, { value }) {
    try {
        // Explicitly handle boolean as viem might require number/bigint/string/bytes
        const valueToConvert = typeof value === "boolean" ? (value ? 1 : 0) : value;
        const hexValue = (0, viem_1.toHex)(valueToConvert); // Cast needed after boolean check
        return `Hex: ${hexValue}`;
    }
    catch (error) {
        return formatError("to_hex", error);
    }
}
class ToHexAction {
    constructor() {
        this.name = "to_hex";
        this.description = exports.ToHexPrompt;
        this.argsSchema = exports.ToHexSchema;
        this.func = toHexFunc;
        this.smartAccountRequired = false;
    }
}
exports.ToHexAction = ToHexAction;
// --- 6. From Hex ---
exports.FromHexSchema = zod_1.z.object({
    hex: zod_1.z.string().refine(viem_1.isHex, "Input must be a valid hex string (0x...)."),
    to: zod_1.z
        .enum(["string", "number", "bigint", "boolean"])
        .describe("The target type to convert the hex value to."),
});
exports.FromHexPrompt = `
Name: from_hex
Description: Converts a hexadecimal string (0x...) into a specified type (string, number, bigint, or boolean).
Usage: Useful for decoding hex values received from the blockchain or other sources.
Input Parameters:
  - hex (string, required): The hexadecimal string (must start with 0x).
  - to (string, required): The target type. Must be one of: "string", "number", "bigint", "boolean".
Output:
  - On success: Returns the decoded value. Example: "Decoded Value: 96" (for "0x60", "number")
  - On failure: Returns an error message.
`;
async function fromHexFunc(_wallet, { hex, to }) {
    try {
        const value = (0, viem_1.fromHex)(hex, to);
        return `Decoded Value: ${value.toString()}`; // Ensure output is stringifiable
    }
    catch (error) {
        return formatError("from_hex", error);
    }
}
class FromHexAction {
    constructor() {
        this.name = "from_hex";
        this.description = exports.FromHexPrompt;
        this.argsSchema = exports.FromHexSchema;
        this.func = fromHexFunc;
        this.smartAccountRequired = false;
    }
}
exports.FromHexAction = FromHexAction;
// --- 7. To Bytes ---
exports.ToBytesSchema = zod_1.z.object({
    value: zod_1.z
        .union([zod_1.z.string(), zod_1.z.number(), zod_1.z.bigint()])
        .describe("The value (string, number, bigint) to convert to a byte array."),
    // opts: z.object({ size: z.number().int().positive().optional(), signed: z.boolean().optional() }).optional().describe("Optional settings for size and signedness.")
});
exports.ToBytesPrompt = `
Name: to_bytes
Description: Converts a string (hex or UTF-8), number, or bigint into a Uint8Array byte representation. The output format is not directly usable by the agent but this tool confirms the conversion process. Primarily useful for internal checks or confirming data before encoding.
Usage: Use when needing to represent data as raw bytes, often as an intermediate step. Output is a description, not the raw bytes.
Input Parameters:
  - value (string | number | bigint, required): The value to convert. Strings are interpreted as hex if they start with 0x, otherwise UTF-8.
Output:
  - On success: Returns a confirmation message indicating the byte array was generated and its length. Example: "Value converted to byte array (Uint8Array) with length: 5"
  - On failure: Returns an error message.
`;
// This function won't return the actual Uint8Array to the agent, as it's not easily serializable/usable.
// It will return a confirmation string instead.
async function toBytesFunc(_wallet, { value }) {
    try {
        let bytes;
        if (typeof value === "string") {
            bytes = (0, viem_1.stringToBytes)(value); // handles hex and utf8 automatically via viem v2+
        }
        else {
            // For number/bigint, need more context (size, signed). For simplicity, using basic toHex -> hexToBytes
            const hexVal = (0, viem_1.toHex)(value);
            bytes = (0, viem_1.hexToBytes)(hexVal);
        }
        // const bytes = toBytes(value as string | number | bigint /*, opts */);
        return `Value converted to byte array (Uint8Array) with length: ${bytes.length}`;
    }
    catch (error) {
        return formatError("to_bytes", error);
    }
}
class ToBytesAction {
    constructor() {
        this.name = "to_bytes";
        this.description = exports.ToBytesPrompt;
        this.argsSchema = exports.ToBytesSchema;
        this.func = toBytesFunc;
        this.smartAccountRequired = false;
    }
}
exports.ToBytesAction = ToBytesAction;
// --- 8. From Bytes ---
exports.FromBytesSchema = zod_1.z.object({
    // Represent bytes as hex for agent input
    hexBytes: zod_1.z
        .string()
        .refine(viem_1.isHex, "Input must be a valid hex string representing bytes (0x...)."),
    to: zod_1.z
        .enum(["string", "number", "bigint", "boolean"])
        .describe("The target type to convert the bytes to."),
    // opts: z.object({ size: z.number().int().positive().optional(), signed: z.boolean().optional() }).optional().describe("Optional settings for size and signedness for number/bigint conversion.")
});
exports.FromBytesPrompt = `
Name: from_bytes
Description: Converts a byte array (represented as a hex string) into a specified type (string, number, bigint, or boolean).
Usage: Use when you have raw byte data (as hex) and need to interpret it as a specific data type.
Input Parameters:
  - hexBytes (string, required): The byte array represented as a hexadecimal string (0x...).
  - to (string, required): The target type. Must be one of: "string", "number", "bigint", "boolean".
Output:
  - On success: Returns the decoded value as a string. Example: "Decoded Value: HelloWorld" (for 'string') or "Decoded Value: 96" (for 'number')
  - On failure: Returns an error message.
`;
async function fromBytesFunc(_wallet, { hexBytes, to /*, opts */ }) {
    try {
        const bytes = (0, viem_1.hexToBytes)(hexBytes);
        let value;
        // viem's fromBytes is mainly for number/bigint with options.
        // For string, use bytesToString. For boolean, check if non-zero.
        switch (to) {
            case "string":
                value = (0, viem_1.bytesToString)(bytes);
                break;
            case "number":
            case "bigint":
            case "boolean": {
                // Use fromHex for simplicity as it handles these conversions
                // Direct fromBytes requires size/signed options not easily managed here.
                const intermediateHex = (0, viem_1.bytesToHex)(bytes);
                value = (0, viem_1.fromHex)(intermediateHex, to); // Reuse fromHex logic
                break;
            }
            default: {
                return formatError("from_bytes", `Unsupported target type: ${to}`);
            }
        }
        //const value = fromBytes(bytes, { to: to as 'string' | 'number' | 'bigint' | 'boolean' /* , ...opts */ });
        return `Decoded Value: ${value.toString()}`;
    }
    catch (error) {
        return formatError("from_bytes", error);
    }
}
class FromBytesAction {
    constructor() {
        this.name = "from_bytes";
        this.description = exports.FromBytesPrompt;
        this.argsSchema = exports.FromBytesSchema;
        this.func = fromBytesFunc;
        this.smartAccountRequired = false;
    }
}
exports.FromBytesAction = FromBytesAction;
// --- Exports ---
// You might want to export these actions individually or as an array/object
exports.formatHelperActions = [
    new FormatUnitsAction(),
    new ParseUnitsAction(),
    new FormatEtherAction(),
    new ParseEtherAction(),
    new ToHexAction(),
    new FromHexAction(),
    new ToBytesAction(),
    new FromBytesAction(),
];
