// @0xgasless/agentkit/dist/BaseActions/SendTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendTransactionAction = exports.SendTransactionPrompt = exports.SendTransactionSchema = void 0;
exports.sendTransactionFunc = sendTransactionFunc;
const zod_1 = require("zod");
const smart_account_1 = require("@0xgasless/smart-account");
exports.SendTransactionSchema = zod_1.z.object({
    to: zod_1.z.string().regex(/^0x[a-fA-F0-9]{40}$/, "Invalid 'to' address format."),
    data: zod_1.z
        .string()
        .regex(/^0x[a-fA-F0-9]*$/, "Invalid 'data' hex format.")
        .optional()
        .nullable(),
    value: zod_1.z
        .string()
        .regex(/^\d+$/, "Invalid 'value' format, should be a string of wei.")
        .optional()
        .nullable(),
});
// 2. Define the prompt detailing the tool's use
exports.SendTransactionPrompt = `
Name: send_transaction
Description: Sends a transaction using the 0xgasless Smart Account. This action submits a transaction to the blockchain to interact with contracts or transfer native tokens.
Usage: Use this action when you need to execute a blockchain transaction, such as calling a contract function or sending the native currency (e.g., ETH, MATIC).
Input Parameters:
  - to (string, required): The destination address for the transaction (e.g., contract address or recipient EOA). Must be a valid Ethereum address (0x...).
  - data (string, optional): The encoded function data for contract interactions. Required when calling a contract function. Must be a hex string (0x...). Use 'encode_function_data' tool to generate this if needed.
  - value (string, optional): The amount of native currency (in wei) to send with the transaction. Defaults to "0" if not provided. Must be a string representing an integer.
Output:
  - On success: Returns a confirmation message including the User Operation Hash (userOpHash). Example: "Transaction submitted successfully! User Operation Hash: 0x..."
  - On failure: Returns an error message detailing the issue. Example: "Error: Failed to send transaction: Invalid 'to' address."

Note: This action only *submits* the transaction. Use the 'get_transaction_status' action with the returned userOpHash to check if the transaction has been confirmed on the blockchain.
`;
// 3. Define the core function logic
async function sendTransactionFunc(wallet, // The wallet instance is passed implicitly by the agent runner
{ to, data, value }) {
    try {
        // Construct the transaction object
        const tx = {
            to: to, // Cast to expected type after validation
            data: data, // Cast to expected type after validation
            value: value ? BigInt(value) : BigInt(0), // Convert validated string to bigint
        };
        // Send the transaction using the 0xgasless SDK
        const request = await wallet.sendTransaction(tx, {
            paymasterServiceData: {
                mode: smart_account_1.PaymasterMode.SPONSORED, // Use sponsored mode for gasless
            },
        });
        // Handle potential errors returned by the SDK
        if (request.error) {
            return `Error: Failed to send transaction: ${request.error.message || request.error}`;
        }
        // Return success message with the userOpHash
        return `Transaction submitted successfully! User Operation Hash: ${request.userOpHash}\nUse 'get_transaction_status' to check confirmation.`;
    }
    catch (error) {
        // Catch any other exceptions during the process
        console.error("Error in sendTransactionFunc:", error);
        return `Error: Failed to send transaction: ${error instanceof Error ? error.message : String(error)}`;
    }
}
class SendTransactionAction {
    constructor() {
        this.name = "send_transaction";
        this.description = exports.SendTransactionPrompt; // Use the prompt string
        this.argsSchema = exports.SendTransactionSchema; // Use the zod schema
        this.func = sendTransactionFunc;
        this.smartAccountRequired = true; // This action needs the smart account wallet
    }
}
exports.SendTransactionAction = SendTransactionAction;
