// @0xgasless/agentkit/dist/BaseActions/SignMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignMessageAction = exports.SignMessagePrompt = exports.SignMessageSchema = void 0;
exports.signMessageFunc = signMessageFunc;
const zod_1 = require("zod");
// 1. Define the schema for the input parameters
exports.SignMessageSchema = zod_1.z.object({
    message: zod_1.z.string().min(1, "Message cannot be empty."),
});
// 2. Define the prompt detailing the tool's use
exports.SignMessagePrompt = `
Name: sign_message
Description: Signs an arbitrary message using the 0xgasless Smart Account's private key.
Usage: Use this action when an application or protocol requires you to prove ownership of your address by signing a specific message.
Input Parameters:
  - message (string, required): The message string that needs to be signed.
Output:
  - On success: Returns the signature as a hexadecimal string (0x...). Example: "Signature: 0x..."
  - On failure: Returns an error message detailing the issue. Example: "Error: Failed to sign message: User rejected signing."
`;
// 3. Define the core function logic
async function signMessageFunc(wallet, // The wallet instance is passed implicitly
{ message }) {
    try {
        // Sign the message using the smart account's underlying signer
        const signature = await wallet.signMessage(message);
        // Return the signature
        return `Signature: ${signature}`;
    }
    catch (error) {
        // Catch any exceptions during the signing process
        console.error("Error in signMessageFunc:", error);
        return `Error: Failed to sign message: ${error instanceof Error ? error.message : String(error)}`;
    }
}
// 4. Define the AgentkitAction class
class SignMessageAction {
    constructor() {
        this.name = "sign_message";
        this.description = exports.SignMessagePrompt;
        this.argsSchema = exports.SignMessageSchema;
        this.func = signMessageFunc;
        this.smartAccountRequired = true; // Signing requires the smart account
    }
}
exports.SignMessageAction = SignMessageAction;
