// @0xgasless/agentkit/dist/actions/checkTransactionAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckTransactionAction = exports.CheckTransactionInput = void 0;
exports.checkTransactionStatus = checkTransactionStatus;
const zod_1 = require("zod");
const services_1 = require("../services");
const CHECK_TRANSACTION_PROMPT = `
This tool checks the status of a previously submitted transaction using its User Operation Hash.
It will attempt to get the transaction receipt and confirmation status.

Required parameters:
- userOpHash: The User Operation Hash returned when the transaction was submitted
- confirmations: (Optional) Number of block confirmations to wait for (default: 1)
- maxDuration: (Optional) Maximum time to wait in milliseconds (default: 30000)
- interval: (Optional) How often to check status in milliseconds (default: 5000)
`;
exports.CheckTransactionInput = zod_1.z
    .object({
    userOpHash: zod_1.z.string().describe("The User Operation Hash to check"),
    confirmations: zod_1.z
        .number()
        .optional()
        .nullable()
        .describe("Number of block confirmations to wait for (default: 1)"),
    maxDuration: zod_1.z
        .number()
        .optional()
        .nullable()
        .describe("Maximum time to wait in milliseconds (default: 30000)"),
    interval: zod_1.z
        .number()
        .optional()
        .nullable()
        .describe("How often to check status in milliseconds (default: 5000)"),
})
    .strip()
    .describe("Instructions for checking transaction status");
/**
 * Checks the status of a transaction using its User Operation Hash.
 *
 * @param wallet - The smart account to use for checking status.
 * @param args - The input arguments containing the userOpHash and optional parameters.
 * @returns A message containing the transaction status.
 */
async function checkTransactionStatus(wallet, args) {
    try {
        const status = await (0, services_1.waitForTransaction)(wallet, args.userOpHash, {
            confirmations: args.confirmations ?? undefined,
            maxDuration: args.maxDuration ?? undefined,
            interval: args.interval ?? undefined,
        });
        switch (status.status) {
            case "confirmed":
                return `
Transaction confirmed!
Block Number: ${status.blockNumber}
Block Confirmations: ${status.blockConfirmations}
Receipt: ${JSON.stringify(status.receipt, null, 2)}
        `;
            case "pending":
                return `
Transaction is still pending.
${status.error ? `Note: ${status.error}` : ""}
You can try checking again with a longer maxDuration.
        `;
            case "failed":
                return `
Transaction failed!
Error: ${status.error}
        `;
            default:
                return `Unknown transaction status`;
        }
    }
    catch (error) {
        return `Error checking transaction status: ${error}`;
    }
}
/**
 * Check transaction status action.
 */
class CheckTransactionAction {
    constructor() {
        this.name = "check_transaction_status";
        this.description = CHECK_TRANSACTION_PROMPT;
        this.argsSchema = exports.CheckTransactionInput;
        this.func = checkTransactionStatus;
        this.smartAccountRequired = true;
    }
}
exports.CheckTransactionAction = CheckTransactionAction;
