// @0xgasless/agentkit/dist/actions/createAndStoreKeyAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateAndStoreKeyAction = exports.CreateAndStoreKeyArgsSchema = exports.CREATE_AND_STORE_KEY_PROMPT = void 0;
const zod_1 = require("zod");
const keyManagementService_1 = require("../services/keyManagementService");
exports.CREATE_AND_STORE_KEY_PROMPT = `
This tool generates a new Ethereum private key and returns the associated wallet address.
If in a Node.js/Bun environment, the key is stored locally in a SQLite database (keys.db);
this does not occur in browser environments. The private key itself is not returned by the action.
`;
exports.CreateAndStoreKeyArgsSchema = zod_1.z.object({});
class CreateAndStoreKeyAction {
    constructor() {
        this.name = "create_and_store_key";
        this.description = exports.CREATE_AND_STORE_KEY_PROMPT;
        this.argsSchema = exports.CreateAndStoreKeyArgsSchema;
        this.smartAccountRequired = false;
    }
    async func(_args) {
        try {
            const privateKey = (0, keyManagementService_1.generateNewViemPrivateKey)();
            const walletAddress = (0, keyManagementService_1.getAddressFromPrivateKey)(privateKey);
            // Attempt to store the key. storeKeyInDb handles environment check.
            const stored = await (0, keyManagementService_1.storeKey)(walletAddress, privateKey);
            if (stored) {
                console.log(`CreateAndStoreKeyAction: Key for ${walletAddress} stored successfully.`);
            }
            else {
                // This can mean it's a browser environment or DB storing failed in Node.
                // The keyManagementService logs details for Node failures.
                console.log(`CreateAndStoreKeyAction: Key for ${walletAddress} not stored to DB (may be browser environment or DB error).`);
            }
            return JSON.stringify({ walletAddress });
        }
        catch (error) {
            console.error("Error in CreateAndStoreKeyAction:", error);
            const errorMessage = error instanceof Error ? error.message : String(error);
            return JSON.stringify({ error: `Failed to create and store key: ${errorMessage}` });
        }
    }
}
exports.CreateAndStoreKeyAction = CreateAndStoreKeyAction;
