// @0xgasless/agentkit/dist/actions/getAddressAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAddressAction = exports.GetAddressInput = void 0;
exports.getAddress = getAddress;
const zod_1 = require("zod");
const GET_ADDRESS_PROMPT = `
This tool retrieves the smart account address that is already configured with the SDK.
No additional wallet setup or private key generation is needed.

USAGE GUIDANCE:
- When a user asks for their wallet address, account address, or smart account address, use this tool immediately
- No parameters are needed to retrieve the address
- The address can be used for receiving tokens or for verification purposes
- This is a read-only operation that doesn't modify any blockchain state

Note: This action works on all supported networks (Base, Fantom, Moonbeam, Metis, Avalanche, BSC).
`;
exports.GetAddressInput = zod_1.z
    .object({})
    .strip()
    .describe("No input required to get the smart account address");
/**
 * Gets the smart account address.
 *
 * @returns A message containing the smart account address.
 */
async function getAddress(wallet, args) {
    try {
        const smartAccount = await wallet.getAddress(args);
        return `Smart Account: ${smartAccount}`;
    }
    catch (error) {
        console.error("Error getting address:", error);
        return `Error getting address: ${error instanceof Error ? error.message : String(error)}`;
    }
}
/**
 * Get smart account address action.
 */
class GetAddressAction {
    constructor() {
        this.name = "get_address";
        this.description = GET_ADDRESS_PROMPT;
        this.argsSchema = exports.GetAddressInput;
        this.func = getAddress;
        this.smartAccountRequired = true;
    }
}
exports.GetAddressAction = GetAddressAction;
