// @0xgasless/agentkit/dist/actions/getTokenDetailsAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetTokenDetailsAction = exports.GetTokenDetailsInput = void 0;
exports.getTokenDetails = getTokenDetails;
const zod_1 = require("zod");
const services_1 = require("../services");
const GET_TOKEN_DETAILS_PROMPT = `
This tool will fetch details about an ERC20 token including:
- Token name
- Token symbol
- Decimals
- Contract address
- Chain ID

Provide the token contract address to get its details.
`;
exports.GetTokenDetailsInput = zod_1.z
    .object({
    tokenAddress: zod_1.z.string().describe("The ERC20 token contract address"),
})
    .strip()
    .describe("Instructions for getting token details");
/**
 * Gets details about an ERC20 token.
 *
 * @param wallet - The smart account to use for querying.
 * @param args - The input arguments containing the token address.
 * @returns A message containing the token details.
 */
async function getTokenDetails(wallet, args) {
    try {
        const details = await (0, services_1.fetchTokenDetails)(wallet, args.tokenAddress);
        if (!details) {
            return "Error getting token details";
        }
        return `
    Token Details:
            Name: ${details.name}
            Symbol: ${details.symbol}
            Decimals: ${details.decimals}
            Address: ${details.address}
            Chain ID: ${details.chainId}
    `;
    }
    catch (error) {
        return `Error getting token details: ${error}`;
    }
}
/**
 * Get token details action.
 */
class GetTokenDetailsAction {
    constructor() {
        this.name = "get_token_details";
        this.description = GET_TOKEN_DETAILS_PROMPT;
        this.argsSchema = exports.GetTokenDetailsInput;
        this.func = getTokenDetails;
        this.smartAccountRequired = true;
    }
}
exports.GetTokenDetailsAction = GetTokenDetailsAction;
