// @0xgasless/agentkit/dist/actions/smartTransferAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartTransferAction = exports.SmartTransferInput = void 0;
exports.smartTransfer = smartTransfer;
const zod_1 = require("zod");
const viem_1 = require("viem");
const constants_1 = require("../constants");
const services_1 = require("../services");
const SMART_TRANSFER_PROMPT = `
This tool will transfer an ERC20 token or native currency from the wallet to another onchain address using gasless transactions.

It takes the following inputs:
- amount: The amount to transfer
- tokenAddress: The token contract address (use 'eth' for native currency transfers)
- destination: Where to send the funds (must be a valid onchain address)

Important notes:
- Gasless transfers are only available on supported networks: Avalanche C-Chain, Metis chain, BASE, BNB chain, FANTOM, Moonbeam.
- The transaction will be submitted and the tool will wait for confirmation by default.
`;
/**
 * Input schema for smart transfer action.
 */
exports.SmartTransferInput = zod_1.z
    .object({
    amount: zod_1.z.string().describe("The amount of tokens to transfer"),
    tokenAddress: zod_1.z
        .string()
        .describe("The token contract address or 'eth' for native currency transfers"),
    destination: zod_1.z.string().describe("The recipient address"),
})
    .strip()
    .describe("Instructions for transferring tokens from a smart account to an onchain address");
/**
 * Transfers assets using gasless transactions.
 *
 * @param wallet - The smart account to transfer from.
 * @param args - The input arguments for the action.
 * @returns A message containing the transfer details.
 */
async function smartTransfer(wallet, args) {
    try {
        const isEth = args.tokenAddress.toLowerCase() === "eth";
        let tx;
        if (isEth) {
            // Native ETH transfer
            tx = {
                to: args.destination,
                data: "0x",
                value: (0, viem_1.parseEther)(args.amount),
            };
        }
        else {
            // ERC20 token transfer
            const decimals = await wallet.rpcProvider.readContract({
                abi: constants_1.TokenABI,
                address: args.tokenAddress,
                functionName: "decimals",
            });
            const data = (0, viem_1.encodeFunctionData)({
                abi: constants_1.TokenABI,
                functionName: "transfer",
                args: [
                    args.destination,
                    (0, viem_1.parseUnits)(args.amount, decimals || 18),
                ],
            });
            tx = {
                to: args.tokenAddress,
                data,
                value: 0n,
            };
        }
        const response = await (0, services_1.sendTransaction)(wallet, tx);
        if (!response || !response.success) {
            return `Transaction failed: ${response?.error || "Unknown error"}`;
        }
        return `The transaction has been confirmed on the blockchain. Successfully transferred ${args.amount} ${isEth ? "ETH" : `tokens from contract ${args.tokenAddress}`} to ${args.destination}. Transaction Hash: ${response.txHash}`;
    }
    catch (error) {
        return `Error transferring the asset: ${error}`;
    }
}
/**
 * Smart transfer action.
 */
class SmartTransferAction {
    constructor() {
        this.name = "smart_transfer";
        this.description = SMART_TRANSFER_PROMPT;
        this.argsSchema = exports.SmartTransferInput;
        this.func = smartTransfer;
        this.smartAccountRequired = true;
    }
}
exports.SmartTransferAction = SmartTransferAction;
